FEAT(FTDA): Setup RTK Query and hookup '/todos' endpoint with the FE to render all todo items

yamin-faruque-sazim opened the pull request:
**DESCRIPTION OF CHANGES**
- Set up Redux store and added RTK Query API slice for todos
- Configured base API endpoint to connect to the backend
- Implemented `useGetTodosQuery` hook to fetch todos from the backend API
- Fixed issue with `dueDate` by converting it to a readable date format
- Updated Task type property from summary to description to match the BE response 
- Successfully rendered and displayed the list of todos fetched from the database


https://github.com/user-attachments/assets/5dd4dc76-34c2-4521-b281-2393cf31ddac

yamin-faruque-sazim committed:
FEAT(FTDA): Setup RTK Query and hookup '/todos' endpoint with the FE to render all todo items

ishmam-iqbal-sazim commented on services/todoApi.ts:
```diff
@@ -0,0 +1,15 @@
+
+export const todoApi = createApi({
+  reducerPath: "todoApi",
+  baseQuery: fetchBaseQuery({ baseUrl: "http://localhost:3001" }),
```
Best practice is to put the url in your `.env` file and import that var from there.

ishmam-iqbal-sazim commented on modules/HomePage/components/TaskList/TaskList.tsx:
```diff
@@ -85,14 +86,15 @@ const TaskList: React.FC<TaskListProps> = ({
   };
-
+  const { data: todos } = useGetTodosQuery();
+  console.log("Todos", todos);
```
Let's remove unnecessary console logs

ishmam-iqbal-sazim commented on modules/HomePage/components/TaskList/TaskList.tsx:
```diff
@@ -85,14 +86,15 @@ const TaskList: React.FC<TaskListProps> = ({
     setEditFormState({ title: "", description: "", dueDate: "", priority: 1 });
   };
-
+  const { data: todos } = useGetTodosQuery();
```
I think we should also handle isLoading from the query and display a loading indicator if its true.

ishmam-iqbal-sazim commented on modules/HomePage/containers/MainContainer.tsx:
```diff
@@ -2,7 +2,7 @@ import { Container, Button, Group } from "@mantine/core";
 import TaskForm from "../components/TaskForm/TaskForm";
 import TaskList from "../components/TaskList/TaskList";
-import useTaskOperations from "@/shared/utils/useTaskOperations";
+import useTaskOperations from "@/shared/utils/usetaskOperations";
```
@ imports should be above relative imports

ishmam-iqbal-sazim commented on modules/TaskDetailsPage/TaskDetails.tsx:
```diff
@@ -13,7 +13,7 @@ import {
 import classes from "./TaskDetails.module.css";
 import { Task } from "../HomePage/types/Task";
-import useTaskOperations from "@/shared/utils/useTaskOperations";
+import useTaskOperations from "@/shared/utils/usetaskOperations";
```
Same import comment as above

yamin-faruque-sazim committed:
FIX(FTDA):
- Remove unnecessary console logs
- Transfer API base url to .env.local file
- Handle isLoading state via mantine
- Fix imports

yamin-faruque-sazim commented:
Addressed PR comments

ishmam-iqbal-sazim commented on .gitignore:
```diff
@@ -61,3 +61,6 @@ jspm_packages
 .next
+
+# Env variables
+.env.local 
```
Let's add new line here

ishmam-iqbal-sazim commented on modules/TaskDetailsPage/TaskDetails.tsx:
```diff
@@ -10,11 +9,12 @@ import {
   TextInput,
   Select,
 } from "@mantine/core";
+import useTaskOperations from "@/shared/utils/usetaskOperations";
```
Should be new line above

ishmam-iqbal-sazim commented on modules/HomePage/components/TaskList/TaskList.tsx:
```diff
@@ -16,6 +17,7 @@ import { useDisclosure } from "@mantine/hooks";
 import { Task } from "../../types/Task";
 import classes from "./TaskList.module.css";
 import { PRIORITY_OPTIONS } from "@/shared/constants/taskContstants";
+import { useGetTodosQuery } from "@/services/todoApi";
```
@ imports should be above relative imports

ishmam-iqbal-sazim commented on modules/HomePage/components/TaskList/TaskList.tsx:
```diff
@@ -85,14 +87,17 @@ const TaskList: React.FC<TaskListProps> = ({
   };
-
+  const { data: todos, isLoading } = useGetTodosQuery();
+  if (isLoading) {
+    return <Loader size="xl" />;
+  }
```
Let's add new lines above and below

ishmam-iqbal-sazim commented on modules/HomePage/components/TaskList/TaskList.tsx:
```diff
@@ -104,7 +109,7 @@ const TaskList: React.FC<TaskListProps> = ({
       </Modal>
 
-      {sortedTasks.map((task) => (
+      {todos?.map((task) => (
```
What if `todos.length === 0`? Wouldn't an error be thrown inside the map loop?

yamin-faruque-sazim replied:
Are you saying, if there are no todo item in the db, I might get an error here with this implementation? How I am understanding is, if todos.length ===0, that means there are no items in that array, ie no todos in the db. So no todos would be rendered.

ishmam-iqbal-sazim reviewed (approved):
Approved pending comments

yamin-faruque-sazim committed:
NIT(FTDA): Fix spacing
